jaxopt/_src/levenberg_marquardt.py

```diff
@@ -0,0 +1,372 @@
+        gain_ratio > 0, gain_ratio_test_true_func, gain_ratio_test_false_func,
+        *gain_ratio_test_init_state)
+
+    return params, damping_factor, increase_factor, gradient, residual
+
+
+  def update_state_using_delta_params(self, loss_curr, params, delta_params,
+                                      damping_factor, increase_factor,
+                                      gradient, residual, *args, **kwargs):
+    """The function to return state variables based on delta_params.
+
+    Define the functions required for the major conditional of the algorithm,
+    which checks the magnitude of dparams and checks if it is small enough.
+    for the value of dparams."""
+
+    updated_params = params + delta_params
+
+    residual_next = self._fun(updated_params, *args, **kwargs)
+
+    # Calculate denominator of the gain ratio based on Eq. 6.16, "Introduction
+    # to optimization and data fitting", L(0)-L(hlm)=0.5*hlm^T*(mu*hlm-g).
+    gain_ratio_denom = 0.5 * jnp.dot(
+      jnp.transpose(delta_params),
+      (damping_factor * delta_params - gradient))
+
+    # Current value of loss function F=0.5*||f||^2.
+    loss_next = 0.5 * jnp.dot(residual_next, residual_next)
+
+    gain_ratio = (loss_curr - loss_next) / gain_ratio_denom
+
+    gain_ratio_test_init_state = (params, damping_factor,
+                                increase_factor, residual, gradient,
+                                updated_params)
+
+    # Calling the jax condition function:
+    # Note that only the parameters that are used in the rest of the program
+    # are returned by this function.
+
+    params, damping_factor, increase_factor, gradient, residual = (
+      self.update_state_using_gain_ratio(gain_ratio,
+                                         gain_ratio_test_init_state, *args,
+                                         **kwargs))
+
+    return params, damping_factor, increase_factor, gradient, residual
+
+  def update(self, params, state: NamedTuple, *args, **kwargs) -> base.OptStep:
+    """Performs one iteration of the least-squares solver.
+
+    Args:
+      params: pytree containing the parameters.
+      state: named tuple containing the solver state.
+
+    Returns:
+      (params, state)
+    """
+
+    # Current value of the loss function F=0.5*||f||^2.
+    loss_curr = state.loss
+
+    matvec = lambda v: self._jtj_op(params, v, *args, **kwargs)
+
+    if isinstance(self.solver, Callable):
+      delta_params = self.solver(
+          matvec, state.gradient, ridge=state.damping_factor, init=state.delta)
+      if self.geodesic:
+        # jtrpp=JT * r" - J is jacobian and r" is second order dir. derivative
+        rpp = self._d2fvv_op(params, delta_params, delta_params, *args, **kwargs)
+        jtrpp = self._jt_op(params, rpp, *args, **kwargs)
+        delta_params += 0.5*self.solver(matvec, jtrpp, ridge=state.damping_factor)
+    elif self.solver == 'cg':
+      delta_params = solve_cg(
```

Comment:
Any reason to keep cg explicitly? Can't we support it via Callable? Especially since I see you don't test this branch anymore.

Reply:
Done. Thanks.